### PR TITLE
feat: MCP L1

### PR DESCRIPTION
**Pull request by tynes**
**Description**

MCP L1 contracts release

### CODE REVIEW

File: packages/contracts-bedrock/snapshots/abi/L1CrossDomainMessenger.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/abi/L1CrossDomainMessenger.json:1]: Just noting that I've reviewed the ABI diff and don't see any breaking or backwards-incompatible changes there. All ABI changes are either:
1. `constructor` signature changes
2. `initialize` signature changes
3. New methods and events

Which should not affect users.

Planning to check the storage layout next

File: packages/contracts-bedrock/src/L1/SystemConfig.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L1/SystemConfig.sol]: We should set the start block slot before calling initialize so that inside of initialize the setting of the start block is skipped. We should add a comment clarifying that its set particularly this way before initialize. Chains deployed before the introduction of the start block need to have the slot set manually and newly deployed chains will get the value set automatically

File: packages/contracts-bedrock/src/libraries/Arithmetic.sol
Comment (tynes) [inline packages/contracts-bedrock/src/libraries/Arithmetic.sol]: why is this version changed? Ideally it can not be changed

File: packages/contracts-bedrock/test/L1/L2OutputOracle.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L1/L2OutputOracle.t.sol:48]: Are you sure that we have test coverage of these tests elsewhere?

Comment (tynes) [reply packages/contracts-bedrock/test/L1/L2OutputOracle.t.sol:48]: Oh i see, the tests where moved to the bottom of the file

File: packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json]: This is being packed into the intializer slot, are we ok with that? It's not a breaking change but I recall discussing if we wanted to avoid this. If we never drop into assembly to modify slot 0 (and I don't think we do) this is safe, so to be clear I'm ok keeping it

File: packages/contracts-bedrock/snapshots/storageLayout/L2ERC721Bridge.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/L2ERC721Bridge.json]: Same comment re packing with initializer data

File: packages/contracts-bedrock/snapshots/storageLayout/OptimismMintableERC20Factory.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/OptimismMintableERC20Factory.json]: same initializer comment

**Review by mds1 — COMMENTED** [review]
No storage layout issues. One question around packing with initializers. I have no problem with, and prefer it, but I recall briefly discussing it as an option for safety a while ago so figured I'd sanity check

File: packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json
Comment (tynes) [reply packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json]: We will need to clear the initialized slot during the 2 step upgrade so we will either need to:
- re-set the messenger in the `initialize` call
- avoid zeroing it out by packing the call to `setStorage` correctly
- putting a spacer in the contract

I am fine with any of these options

**Comment by tynes** [comment]
Looks like this needs a rebase + gas snapshot regeneration before we can merge it

**Comment by 0xfuturistic** [comment]
rebased with [develop](https://github.com/ethereum-optimism/optimism/tree/develop)

File: packages/contracts-bedrock/snapshots/storageLayout/L2ERC721Bridge.json
Comment (tynes) [inline packages/contracts-bedrock/snapshots/storageLayout/L2ERC721Bridge.json]: @mds1 do you think this should be a `bytes30`?

Comment (mds1) [reply packages/contracts-bedrock/snapshots/storageLayout/L2ERC721Bridge.json]: I agree with that. From re-reading [Storage Layout Management](https://www.notion.so/oplabs/Storage-Layout-Management-9fd6aa22b3cb4702b32fe65ae77ff21d?pvs=4#9b983922d0404316883968a09a3154dd) I don't think we ever came to consensus on a spacer format, but my suggestion would be to always use `bytesN` where `N` is the number of bytes needed to fill the rest of the slot without any gap. 

This is because if we later added a `bool` after this spacer, it would get packed into slot 0, and we're relying on the size of `messenger` to avoid packing.

File: packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/L1ERC721Bridge.json]: Same comment on `bytes30` type here

File: packages/contracts-bedrock/snapshots/storageLayout/OptimismMintableERC20Factory.json
Comment (mds1) [inline packages/contracts-bedrock/snapshots/storageLayout/OptimismMintableERC20Factory.json]: same `bytes30` suggestion again

File: packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol
Comment (maurelian) [inline packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:25]: @tynes I'm assuming that you discussed whether or not this is a change to the security model and decided no?

Comment (tynes) [reply packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:25]: We did talk about this, it should be fine to be a minor version bump everywhere

**Comment by tynes** [comment]
Do not merge until https://github.com/ethereum-optimism/optimism/pull/9109 is merged first

**Review by maurelian — APPROVED** [review]
Approving on the basis of previous reviews on this branch.

**Review by mds1 — APPROVED** [review]

**Review by protolambda — APPROVED** [review]
The Go changes look valid to me, but someone with more MCP L1 context should review the contract changes.

**Comment by tynes** [comment]
> The Go changes look valid to me, but someone with more MCP L1 context should review the contract changes.

@maurelian @mds1 and @tynes have reviewed the contract changes

**Review by mslipper — APPROVED** [review]

**Review by hamdiallam — APPROVED** [review]

**Review by roninjin10 — APPROVED** [review]

**Pull request merged**
